### DISCUSSION TITLE
Webpack 5 not removing unused code

### DISCUSSION THREAD

**Discussion by liamdebeasi** [webpack/webpack#14962] [Q&A] [answered]
Hi everyone,

I am running into a situation where Webpack is not properly removing unused code from an app, and I am having a hard time figuring out why. The issue seems to be that doing a re-assignment is causing the issue (Example: https://github.com/liamdebeasi/webpack-reassignment-demo/blob/main/src/components/avatar.js#L9). Is anyone able to shed some light on why a re-assignment would cause Webpack to pull in this additional code?

Minimal reproduction: https://github.com/liamdebeasi/webpack-reassignment-demo

**Steps to reproduce**

1. Run `npm install`.
2. Open `src/index.js`. Observe that only `Badge` is imported.
2. Run `npx webpack`.
3. Run `grep -o -i -E 'avatar bundle|badge bundle' dist/main.js`. This will search the built output for the `"avatar bundle"` and `"badge bundle"` strings. These strings are found in the `src/components/avatar.js` and `src/components/badge.js` files and are used to quickly tell if Webpack is bundling the components when building. Observe that both strings appear, meaning both `Badge` and `Avatar` are bundled with the app.
4. In `components/avatar.js` remove `Avatar = Avatar`.
5. In `components/badge.js` remove `Badge = Badge`.
6. Run `npx webpack`.
7. Run `grep -o -i -E 'avatar bundle|badge bundle' dist/main.js`. Observe that only the ``"badge bundle"`` string appears, meaning only `Badge` is bundled.

**Comment by alexander-akait** [comment]
I think your repo is private due 404

**Reply by liamdebeasi** [reply]
Whoops, fixed. Thank you!

**Comment by alexander-akait** [comment, marked as answer]
Yep, not implemented, why do not use `wrapperFn(class extends HTMLElement { })`, I am afraid other bundlers can do not support it too

**Reply by liamdebeasi** [reply]
We could do `wrapperFn(class extends HTMLElement { })`, but I guess I am interested in understanding why the current usage does not work with Webpack.

I tested this in Rollup, and Rollup was able to properly bundle on the Badge component even with the re-assignment: https://github.com/liamdebeasi/webpack-reassignment-demo/tree/rollup

(Run `npx rollup -c` instead of `npx webpack`)

**Reply by alexander-akait** [reply]
We just do not have code for handling this case... nothing more

**Reply by liamdebeasi** [reply]
Ok makes sense, thank you! Do you think this is something that Webpack could handle in the future? Or is it outside the scope of what Webpack should account for?

**Reply by alexander-akait** [reply]
I think we can improve this, feel free to open an issue

**Reply by liamdebeasi** [reply]
Thanks for all the help. I filed an issue here: https://github.com/webpack/webpack/issues/14963.
